### PR TITLE
refactor: use TLS library v4

### PR DESCRIPTION
**Pull request by gruyaume**
# Description

Replace the TLS library V3 with V4.

We also leverage this opportunity to stop storing the self signed cert and key in Juju secrets as we are already storing those in the workload containers. There is no need for those to be stored in two independent places.

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/unit/test_certificate_helpers.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+from certificates_helpers import certificate_issuer_has_common_name, generate_certificate
+from cryptography import x509
+
+
+def test_given_sans_ip_when_generate_certificate_then_certificate_contain_expected_attributes():
```
Comment (kayra1) [inline tests/unit/test_certificate_helpers.py:8]:
```suggestion
def test_given_sans_ip_when_generate_certificate_then_certificate_contains_expected_attributes():
```

File: src/certificates_helpers.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        x509.CertificateBuilder()
+        .subject_name(subject)
+        .issuer_name(ca_cert_obj.issuer)
+        .public_key(private_key_obj.public_key())
+        .serial_number(x509.random_serial_number())
+        .not_valid_before(datetime.now(timezone.utc))
+        .not_valid_after(datetime.now(timezone.utc) + timedelta(days=validity))
+        .add_extension(
+            x509.SubjectAlternativeName(
+                [x509.IPAddress(ipaddress.ip_address(san)) for san in sans_ips]
```
Comment (kayra1) [inline src/certificates_helpers.py:105]: Should we add the common name as DNS here too? That seems to be standard now.

File: src/charm.py [hunk 3]
```diff
@@ -83,7 +66,7 @@ def __init__(self, framework: ops.Framework):
             for event in [
                 self.on["gocert"].pebble_ready,
                 self.on["gocert"].pebble_custom_notice,
-                self.tls.on.certificate_creation_request,
+                self.on.certificates_relation_changed,
```
Comment (kayra1) [inline src/charm.py:69]: I'd like to keep the formats here the same.
```suggestion
                self.on["certificates"].relation_changed,
```

**Review by kayra1 — APPROVED** [review]
Small nits, thanks

File: src/certificates_helpers.py (hunk 2 above)
Comment (gruyaume) [reply src/certificates_helpers.py:109]: Done, added an argument to pass in the sans_dns

**Review by kayra1 — APPROVED** [review]

**Pull request merged**